app/lib/clients/vault/oidc.rb

```diff
@@ -0,0 +1,151 @@
+user tries to auth, the client connects to the provider, which opens
+up a browser window allowing the user to enter his username/password.
+
+On success, the provider returns an OIDC token, which includes the
+user's email addr.
+
+The client has been configured to map that email address to an entity
+in vault, which has the policy which we want the user to have.
+
+So the mapping goes from the email address on the provider, to the
+policy in vault.  The email addr may not be the best mapping to use.
+Some other piece of user info may ultimately be better used to map the
+user to the policy.  But we don't yet have a good understanding of
+different OIDC provider configurations, so this should be good enough
+for now
+
+Note that this provider is only meant to be used in our dev/test
+environment to excercise the client.  In a prod env, a real OIDC
+provider is configured in.
+
+=end
+module Clients
+  class Vault
+    module Oidc
+      def configure_oidc_provider
+        create_provider_webapp
+        create_provider_with_email_scope
+        create_entity_for_initial_user
+        create_userpass_for_initial_user
+        map_userpass_to_entity
+      end
+
+      def configure_oidc_client(issuer, client_id, client_secret)
+        create_client_config(issuer, client_id, client_secret)
+        create_default_policy_for_role
+        create_default_role(client_id)
+      end
+
+      def configure_oidc_user(name, email, policy)
+        client.sys.put_policy(email, policy)
+        put_entity(name, email)
+        put_entity_alias(name, email, "oidc")
+      end
+
+      private
+      cattr_accessor :client_id
+      cattr_accessor :client_secret
+      WEBAPP_NAME = "identity/oidc/client/astral"
+
+      def oidc_provider
+        ::Vault::Client.new(
+          address: "http://oidc_provider:8300",
+          token: token
+        )
+      end
+
+      def create_provider_webapp
+        oidc_provider.logical.write(
+          WEBAPP_NAME,
+          redirect_uris: get_redirect_uris,
+          assignments: "allow_all")
+        app = oidc_provider.logical.read(WEBAPP_NAME)
+        @@client_id = app.data[:client_id]
+        @@client_secret = app.data[:client_secret]
+      end
+
+      def create_provider_with_email_scope
+        oidc_provider.logical.write("identity/oidc/scope/email",
+                                    template: '{"email": {{identity.entity.metadata.email}}}')
+        oidc_provider.logical.write("identity/oidc/provider/astral",
+                                    issuer: "http://oidc_provider:8300",
+                                    allowed_client_ids: @@client_id,
+                                    scopes_supported: "email")
+      end
+
+      def create_entity_for_initial_user
+        oidc_provider.logical.write("identity/entity",
+                                    policies: "default",
+                                    name: Config[:initial_user][:name],
```

Comment:
This chained lookup with explode if the first key is blank, maybe wrap it in a method and do some validation? Might make these methods a little tidier.

Reply:
so is it standard in ruby to never use chained lookups?

Reply:
done